File: library/arch/src/main/java/com/github/lany192/arch/items/ItemsFragment.java
package com.github.lany192.arch.items;

import android.view.View;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import com.chad.library.adapter4.QuickAdapterHelper;
import com.chad.library.adapter4.layoutmanager.QuickGridLayoutManager;
import com.chad.library.adapter4.loadState.LoadState;
import com.chad.library.adapter4.loadState.trailing.TrailingLoadStateAdapter;
import com.github.lany192.arch.R;
import com.github.lany192.arch.adapter.MultiAdapter;
import com.github.lany192.arch.adapter.ItemBinder;
import com.github.lany192.arch.fragment.VMVBFragment;
import com.github.lany192.arch.utils.ListUtils;
import com.github.lany192.view.DefaultView;
import com.scwang.smart.refresh.layout.SmartRefreshLayout;

import java.util.List;

public abstract class ItemsFragment<VM extends ItemsViewModel, VB extends ViewBinding>
        extends VMVBFragment<VM, VB> {
    private final MultiAdapter mAdapter = new MultiAdapter();
    private QuickAdapterHelper adapterHelper;

    public RecyclerView.LayoutManager getLayoutManager() {
        QuickGridLayoutManager layoutManager = new QuickGridLayoutManager(getContext(), getSpanCount());
        layoutManager.setOrientation(GridLayoutManager.VERTICAL);
        return layoutManager;
    }

    public <T, B extends ViewBinding> void register(ItemBinder<T, B> binder) {
        mAdapter.addBinder(binder);
    }

    public int getSpanCount() {
        return 2;
    }

    public RecyclerView.ItemDecoration getItemDecoration(RecyclerView.Adapter<?> adapter) {
        return null;
    }

    public abstract SmartRefreshLayout getRefreshLayout();

    public abstract RecyclerView getRecyclerView();

    @CallSuper
    @Override
    public void init() {
        super.init();
        adapterHelper = new QuickAdapterHelper.Builder(mAdapter)
                .setTrailingLoadStateAdapter(new TrailingLoadStateAdapter.OnTrailingListener() {

                    @Override
                    public void onLoad() {
                        if (viewModel.loadMoreEnable()) {
                            viewModel.onLoadMore();
                        }
                    }

                    @Override
                    public void onFailRetry() {
                        if (viewModel.loadMoreEnable()) {
                            viewModel.onLoadMore();
                        }
                    }
                })
                .build();
        getRecyclerView().setLayoutManager(getLayoutManager());
        getRecyclerView().setAdapter(mAdapter);
        if (getRecyclerView().getItemDecorationCount() < 1 && getItemDecoration(mAdapter) != null) {
            getRecyclerView().addItemDecoration(getItemDecoration(mAdapter));
        }
        getRefreshLayout().setEnableLoadMore(false);
        getRefreshLayout().setEnableRefresh(viewModel.refreshEnable());
        if (viewModel.refreshEnable()) {
            getRefreshLayout().setOnRefreshListener(refreshLayout -> viewModel.onRefresh());
        }
        //列表状态观察
        viewModel.getListState().observe(this, state -> {
            switch (state) {
                case ERROR:
                    getRefreshLayout().finishRefresh();
                    adapterHelper.setTrailingLoadState(new LoadState.Error(new Throwable("加载失败")));
                    break;
                case REFRESHING:
                    break;
                case REFRESH_FINISH:
                    getRefreshLayout().finishRefresh();
                    break;
                case MORE_LOADING:
                    break;
                case MORE_LOAD_END:
                    getRefreshLayout().finishRefresh();
                    adapterHelper.setTrailingLoadState(new LoadState.NotLoading(true));
                    break;
                case MORE_LOAD_FINISH:
                    getRefreshLayout().finishRefresh();
                    adapterHelper.setTrailingLoadState(new LoadState.NotLoading(true));
                    break;
            }
        });
        viewModel.getItems().observe(this, data -> {
            setList(data.getItems());
            if (ListUtils.isEmpty(data.getItems())) {
                mAdapter.setStateView(getEmptyView());
            }
        });
    }

    public void setList(List<Object> items) {
        mAdapter.setItems(items);
    }

    @NonNull
    public View getEmptyView() {
        DefaultView view = new DefaultView(this);
        view.setImage(R.drawable.default_empty);
        view.setMessage(R.string.default_empty);
        view.setOnRetryListener(() -> viewModel.onLazyLoad());
        return view;
    }

    @NonNull
    public View getNetworkView() {
        DefaultView view = new DefaultView(this);
        view.setImage(R.drawable.default_network);
        view.setMessage(R.string.default_network);
        view.setOnRetryListener(() -> viewModel.onLazyLoad());
        return view;
    }

    @NonNull
    public View getErrorView() {
        DefaultView view = new DefaultView(this);
        view.setImage(R.drawable.default_error);
        view.setMessage(R.string.default_error);
        view.setOnRetryListener(() -> viewModel.onLazyLoad());
        return view;
    }
}

File: library/arch/src/main/java/com/github/lany192/arch/items/ItemsFragment.kt
package com.github.lany192.arch.items

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.chad.library.adapter4.QuickAdapterHelper
import com.chad.library.adapter4.layoutmanager.QuickGridLayoutManager
import com.github.lany192.arch.adapter.MultiAdapter
import com.github.lany192.arch.fragment.VMVBFragment
import com.github.lany192.arch.utils.ListUtils
import com.scwang.smart.refresh.layout.SmartRefreshLayout

abstract class ItemsFragment<VM : ItemsViewModel, VB : ViewBinding> : VMVBFragment<VM, VB>() {
    abstract fun getRefreshLayout(): SmartRefreshLayout

    abstract fun getRecyclerView(): RecyclerView

    protected lateinit var layoutManager: RecyclerView.LayoutManager

    protected lateinit var recyclerView: RecyclerView

    private val mAdapter by lazy(LazyThreadSafetyMode.NONE) {
        MultiAdapter(mutableListOf())
    }

    private val helper by lazy(LazyThreadSafetyMode.NONE) {
        QuickAdapterHelper.Builder(mAdapter).build()
    }

    fun <T : Any, VB : ViewBinding> register(binder: ItemBinder<T, VB>) {
        mAdapter.register(binder)
    }

    open fun createLayoutManager(): RecyclerView.LayoutManager {
        val layoutManager = QuickGridLayoutManager(requireContext(), getSpanCount())
        layoutManager.setOrientation(GridLayoutManager.VERTICAL)
        return layoutManager
    }

    fun addOnScrollListener(listener: RecyclerView.OnScrollListener) {
        recyclerView.addOnScrollListener(listener)
    }

    open fun getSpanCount(): Int {
        return 2
    }

    override fun init() {
        super.init()
        recyclerView = getRecyclerView()
        layoutManager = createLayoutManager()
        recyclerView.setLayoutManager(layoutManager)
        recyclerView.adapter = helper.adapter
        getRefreshLayout().setEnableLoadMore(true)
        getRefreshLayout().setOnLoadMoreListener {
            viewModel.onLoadMore()
        }
        getRefreshLayout().setEnableRefresh(viewModel.refreshEnable())
        if (viewModel.refreshEnable()) {
            getRefreshLayout().setOnRefreshListener {
                viewModel.onRefresh()
            }
        }
        //列表状态观察
        viewModel.listState.observe(this) {
            when (it) {//常量
                ListState.ERROR -> {
                    getRefreshLayout().finishRefresh()
                }

                ListState.REFRESHING -> {

                }

                ListState.REFRESH_FINISH -> {
                    getRefreshLayout().finishRefresh()
                }

                ListState.MORE_LOADING -> {

                }

                ListState.MORE_LOAD_END -> {
                    getRefreshLayout().finishRefresh()
                    getRefreshLayout().finishLoadMore()
                }

                ListState.MORE_LOAD_FINISH -> {
                    getRefreshLayout().finishRefresh()
                    getRefreshLayout().finishLoadMore()
                }
            }
        }
        viewModel.items.observe(this) {
            mAdapter.submitList(it.items)
            if (ListUtils.isEmpty(it.items)) {
                mAdapter.stateView = getEmptyView()
            }
        }
    }
}